circe/src/main/scala/org/http4s/circe/JsonDecoder.scala

```diff
@@ -0,0 +1,25 @@
+package org.http4s.circe
+
+import cats.effect.Sync
+import org.http4s._
+import io.circe._
+
+/**
+  * F-algebra for separating the Sync required for extracting
+  * the Json from the body. As such if F is Sync at some layer,
+  * then this can be used to extract without the lower layer
+  * needing to be aware of the strong constraint.
+ **/
+trait JsonDecoder[F[_]] {
+  def asJson(m: Message[F]): F[Json]
+  def asJsonDecode[A: Decoder](m: Message[F]): F[A]
+}
+
+object JsonDecoder {
+  def apply[F[_]](implicit ev: JsonDecoder[F]): JsonDecoder[F] = ev
+
+  implicit def impl[F[_]: Sync]: JsonDecoder[F] = new JsonDecoder[F] {
```

Comment:
Thinking about what other implementations could plausibly exist.  I guess for the other `Stream.Compiler`s, though you'd want a `MonadError` for `Id` and `Pure`.  It's not a typeclass with laws, so I guess a constant decoder "works".

I get the aversion to `Sync`, but I need some help thinking about why it's good to pretend that this can't fail.